### PR TITLE
masync: add dml mover implementation

### PR DESCRIPTION
**Pull request by kswiecicki**
Includes https://github.com/pmem/miniasync/pull/15.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/miniasync/22)

### CODE REVIEW

**Review by pbalcer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/22)* status: 0 of 13 files reviewed, 4 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[extras/dml/dml_mover.c, line 1 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-Mpzugy80uBHXt33p53f:-Mpzugy80uBHXt33p53g:b-713ynl) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L1)):*
> ```C
> #include <assert.h>
> ```

Reverse the name (mover_dml/vdm_dml). Having our own APIs prefixed with "dml_" is confusing and might actually conflict with DML header.
___
*[extras/dml/dml_mover.c, line 20 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzvW_F-X8kd4tBxFrI:-MpzvW_F-X8kd4tBxFrJ:b-l198py) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L20)):*
> ```C
> 	assert(status == DML_STATUS_OK);
> 
> 	dml_job = (dml_job_t *)malloc(job_size);
> ```

How much memory is this allocating? The last time when Juno worked on this, the value here was huge (~100kb).
___
*[extras/dml/dml_mover.c, line 85 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzuRBCEYXSORf_GSUO:-MpzuRBCEYXSORf_GSUP:bd6wn4x) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L85)):*
> ```C
> 	if (state == FUTURE_STATE_COMPLETE)
> 		dml_finalize_job(dml_job);
> ```

`dml_memcpy_job_delete` ?
___
*[extras/dml/dml_mover.c, line 130 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpztLV4C5gTWNnzroMK:-MpztLV4C5gTWNnzroML:bv9u36o) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L130)):*
> ```C
> static struct vdm_descriptor dml_asynchronous_descriptor = {
> 	.vdm_data = NULL,
> 	.memcpy = dml_memcopy_async,
> ```

a small nit, but very jarring for me: `memcpy` ;)

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 1 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-Mpzugy80uBHXt33p53f:-Mq00N5u5qkE-owleC8I:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L1)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Reverse the name (mover_dml/vdm_dml). Having our own APIs prefixed with "dml_" is confusing and might actually conflict with DML header.
</blockquote></details>

Done.

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 20 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzvW_F-X8kd4tBxFrI:-Mq00O6HAoGZQj1L-3VT:bo576r2) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L20)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

How much memory is this allocating? The last time when Juno worked on this, the value here was huge (~100kb).
</blockquote></details>

I get the value of ```job_size``` as '98816'.

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 85 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzuRBCEYXSORf_GSUO:-Mq00YWx-YcFvvACQu72:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L85)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

`dml_memcpy_job_delete` ?
</blockquote></details>

Done.

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 130 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpztLV4C5gTWNnzroMK:-Mq00ZXu2DTGvg96zIc3:b-kbvo2b) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L130)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

a small nit, but very jarring for me: `memcpy` ;)
</blockquote></details>

I was doing it unconsciously :x

**Review by pbalcer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/22)* status: 0 of 13 files reviewed, all discussions resolved (waiting on @pbalcer)

___
*[extras/dml/dml_mover.c, line 20 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzvW_F-X8kd4tBxFrI:-MqDu9P48Sktf-Pjcj9X:bmc6yte) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L20)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

I get the value of ```job_size``` as '98816'.
</blockquote></details>

That's unfortunate - can you please report a bug to intel/dml?

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 20 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzvW_F-X8kd4tBxFrI:-MqE1K9c0k3v8fw4Yn01:b-80tlqk) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L20)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

That's unfortunate - can you please report a bug to intel/dml?
</blockquote></details>

Sure.

**Comment by kswiecicki** [comment]
___
*[extras/dml/dml_mover.c, line 20 at r3](https://reviewable.io/reviews/pmem/miniasync/22#-MpzvW_F-X8kd4tBxFrI:-MqEBrua3nTClewsXFBd:b-nmmb4r) ([raw file](https://github.com/pmem/miniasync/blob/c5132a332eb0ad5fc4846331308ddce015d8e6ac/extras/dml/dml_mover.c#L20)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Sure.
</blockquote></details>

Created issue: https://github.com/intel/DML/issues/12.

File: extras/dml/vdm_dml.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright 2021, Intel Corporation */
+
+#include <assert.h>
+#include <dml/dml.h>
+#include <libminiasync.h>
+#include <stdatomic.h>
```
Comment (DamianDuy) [inline extras/dml/vdm_dml.c:7]: Perhaps you could replace the atomic_* functions with the wrappers from ```core/util.h``` file?

**Comment by kswiecicki** [comment]
___
*[extras/dml/vdm_dml.c, line 7 at r17](https://reviewable.io/reviews/pmem/miniasync/22#-Mq-zS5g-15KDb6bN_zN-r17-7:-MrMKnMNA81W2vCQn4I8:b-madkjj) ([raw file](https://github.com/pmem/miniasync/blob/b8607d5be99b162117af6afbf3827d088fe57591/extras/dml/vdm_dml.c#L7)):*
<details><summary><i>Previously, DamianDuy (Damian Duy) wrote…</i></summary><blockquote>

Perhaps you could replace the atomic_* functions with the wrappers from ```core/util.h``` file?
</blockquote></details>

Sure, I was waiting for this patch actually.
Done.

File: extras/dml/vdm_dml.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+vdm_dml_check_delete_job(struct future_context *context)
+{
+	struct vdm_memcpy_data *data = future_context_get_data(context);
+	dml_job_t *dml_job = (dml_job_t *)data->extra;
+
+	dml_status_t status = dml_check_job(dml_job);
+	assert(status != DML_STATUS_JOB_CORRUPTED);
+
+	enum future_state state = (status == DML_STATUS_OK) ?
+			FUTURE_STATE_COMPLETE : FUTURE_STATE_RUNNING;
```
Comment (DamianDuy) [inline extras/dml/vdm_dml.c:133]: I'd like to ask, what status the ```dml_check_job``` function is supposed to return for the future state to be set to running?  From what I've looked up in the ```dml.h``` this function returns STATUS_OK, STATUS_JOB_CORRUPTED or STATUS_NULL_POINTER_ERROR. That's why I was wondering if the STATUS_NULL_POINTER_ERROR is used here and if so then why?

Comment (DamianDuy) [reply extras/dml/vdm_dml.c:133]: Never mind. I think I looked up the wrong function.

**Review by lplewa — APPROVED** [review]
:lgtm:

> Reviewed 2 of 13 files at r19, 1 of 8 files at r20, 2 of 3 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/22#-:-MrSjtJDB9xLBBx1TsjQ:bnfp4nl)* status: 5 of 14 files reviewed, all discussions resolved (waiting on @DamianDuy, @lplewa, and @pbalcer)

**Review by DamianDuy — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/22#-:-MrWAYVe60Q_1r843hxE:bnfp4nl)* status: 5 of 14 files reviewed, all discussions resolved (waiting on @DamianDuy, @lplewa, and @pbalcer)

File: CMakeLists.txt [hunk 3]
```diff
@@ -31,6 +31,7 @@ option(BUILD_DOC "build documentation" ON)
 option(BUILD_EXAMPLES "build examples" ON)
 option(BUILD_TESTS "build tests" ON)
 option(TESTS_USE_VALGRIND "enable tests with valgrind (if found)" ON)
+option(COMPILE_DML "compile miniasync dml implementation library" OFF)
```
Comment (wlemkows) [inline CMakeLists.txt:34]: Can it be enabled on gha? We need to test it on build system.

**Comment by kswiecicki** [comment]
___
*[CMakeLists.txt, line 34 at r22](https://reviewable.io/reviews/pmem/miniasync/22#-MpzsvQ_cP3KdMzzjw7A-r22-34:-MrWJNLc7GtVqepvYkO4:bd9jfl) ([raw file](https://github.com/pmem/miniasync/blob/ecd14d9721c83e104ef09966137de83272701f70/CMakeLists.txt#L34)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Can it be enabled on gha? We need to test it on build system.
</blockquote></details>

I've added -DCOMPILE_DML=1 to the build.sh.
Done.

**Comment by kswiecicki** [comment]
___
*[CMakeLists.txt, line 34 at r22](https://reviewable.io/reviews/pmem/miniasync/22#-MpzsvQ_cP3KdMzzjw7A-r22-34:-MrWJX-wAWO2uAu0VgMI:b-59n4qa) ([raw file](https://github.com/pmem/miniasync/blob/ecd14d9721c83e104ef09966137de83272701f70/CMakeLists.txt#L34)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

I've added -DCOMPILE_DML=1 to the build.sh.
Done.
</blockquote></details>

run-build.sh*

File: extras/dml/include/libminiasync-dml/vdm_dml.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#define VDM_DML_H 1
+
+#include <dml/dml.h>
+#include <libminiasync.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct vdm_descriptor *vdm_descriptor_dml(void);
```
Comment (wlemkows) [inline extras/dml/include/libminiasync-dml/vdm_dml.h:14]: documentation is missing

File: extras/dml/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	PUBLIC_HEADER ${MINIASYNC_DML_INCLUDE_DIR}/libminiasync-dml.h
+)
+
+install(TARGETS miniasync-dml
+	PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+	RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+# Install headers included in public header
+install(DIRECTORY ${MINIASYNC_DML_INCLUDE_DIR}/libminiasync-dml
```
Comment (wlemkows) [inline extras/dml/CMakeLists.txt:36]: what about packages?

**Review by wlemkows — CHANGES_REQUESTED** [review]

File: extras/dml/include/libminiasync-dml/vdm_dml.h (hunk 4 above)
Comment (wlemkows) [inline extras/dml/include/libminiasync-dml/vdm_dml.h:14]: Do we have any test??

**Comment by kswiecicki** [comment]
___
*[extras/dml/CMakeLists.txt, line 36 at r27](https://reviewable.io/reviews/pmem/miniasync/22#-MpzsvQ_cP3KdMzzjw7C-r27-36:-MrWlnFqEsNLn7FgV0I6:b-4mxqag) ([raw file](https://github.com/pmem/miniasync/blob/faf0a9548d1f1e4502f2763c1b852d40ed949d85/extras/dml/CMakeLists.txt#L36)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

what about packages? 
</blockquote></details>

Packages for libminiasync-dml are working, it is built into a package whenever COMPILE_DML is ON.

**Comment by kswiecicki** [comment]
___
*[extras/dml/include/libminiasync-dml/vdm_dml.h, line 14 at r27](https://reviewable.io/reviews/pmem/miniasync/22#-Mq-zS5g-15KDb6bN_zM-r27-14:-MrWnMqL2n3TUA7byQOO:b-3d2i8z) ([raw file](https://github.com/pmem/miniasync/blob/faf0a9548d1f1e4502f2763c1b852d40ed949d85/extras/dml/include/libminiasync-dml/vdm_dml.h#L14)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Do we have any test??
</blockquote></details>

Mind if I add documentation and tests in separate PR?

**Pull request merged**